### PR TITLE
Add doc about resolving trim warnings

### PR DESCRIPTION
**Pull request by agocke**
## Summary

Should serve as an introduction to trim warnings and guidance about how to fix the most common problems.

### CODE REVIEW

File: docs/core/deploying/fixing-warnings.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
+---
+title: Introduction to trim warnings
+description: Learn about why warnings may be produced when publishing a trimmed application, how to address them, and how to make the application "trim compatible."
+author: agocke
+ms.author: angocke
+ms.date: 09/08/2021
+---
+# Introduction to trim warnings
+
+Conceptually, trimming is very simple: when publishing the application, the .NET SDK analyzes the
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:10]: These should be single lines without the column cut off

File: docs/core/deploying/fixing-warnings.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+In this case, the developer would expect a warning on the call to `Type.GetType`, indicating that
+it cannot determine which type is going to be used by the pplication.
+
+## Reacting to trim warnings
+
+Trim warnings are meant to bring predictability to trimming. There are two big categories of warnings which you will likely see:
+
+ 1. `RequiresUnreferencedCode`
+ 2. `DynamicallyAccessedMembers`
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:55]: For some reason these are indented
```suggestion
1. `RequiresUnreferencedCode`
2. `DynamicallyAccessedMembers`
```

File: docs/core/deploying/fixing-warnings.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+```C#
+string s = Console.ReadLine();
+Type type = Type.GetType(s);
+foreach (var m in type.GetMethods())
+{
+    Console.WriteLine(m.Name);
+}
+```
+
+In this example, `Type.GetType` dynamically requests a type with an unknown name, and then prints
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:40]: The first introduction of a type should use the xref syntax

File: docs/core/deploying/fixing-warnings.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+ 2. `DynamicallyAccessedMembers`
+
+### RequiresUnreferencedCode
+
+<xref:System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute>
+is simple and broad: it's an attribute that means the member has been annotated incompatible with trimming, meaning
+that it might use reflection or some other mechanism to access code that may be trimmed away. This
+attribute is used when code is fundamentally not trim compatible, or the trim dependency is too complex
+to explain to the trimmer. This would often be true for methods which use the C# `dynamic` keyword,
+accessing types from `Assembly.LoadFrom`, or other runtime code generation technologies.  An
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:64]: xref here for Assembly.LoadFrom

File: docs/core/deploying/fixing-warnings.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+ 1. `RequiresUnreferencedCode`
+ 2. `DynamicallyAccessedMembers`
+
+### RequiresUnreferencedCode
+
+<xref:System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute>
+is simple and broad: it's an attribute that means the member has been annotated incompatible with trimming, meaning
+that it might use reflection or some other mechanism to access code that may be trimmed away. This
+attribute is used when code is fundamentally not trim compatible, or the trim dependency is too complex
+to explain to the trimmer. This would often be true for methods which use the C# `dynamic` keyword,
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:63]: dynamic could link to the csharp article

File: docs/core/deploying/fixing-warnings.md (hunk 4 above)
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:64]: Search for double spaces and remove them.
```suggestion
accessing types from `Assembly.LoadFrom`, or other runtime code generation technologies. An
```

File: docs/core/deploying/fixing-warnings.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+compatible. Adding `RequiresUnreferencedCode` will silence all trimming warnings in the given
+method, but will produce a warning whenever someone else calls it. For this reason, it is mostly
+useful to library authors to "bubble up" the warning to a public API.
+
+If you can somehow determine that the call is safe, and all the code that's needed won't be
+trimmed away, you can also suppress the warning using
+<xref:System.Diagnostics.CodeAnalysis.UnconditionalSuppressMessageAttribute>.
+For example:
+
+```C#
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:92]: We use csharp as the language identifier
```suggestion
```csharp
```

File: docs/core/deploying/fixing-warnings.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+If you compile the above, now you see a warning:
+
+```
+Trim analysis warning IL2070: net6.Program.Method2(Type): 'this' argument does not satisfy
+'DynamicallyAccessedMemberTypes.PublicMethods' in call to 'System.Type.GetMethods()'. The
+parameter 'type' of method 'net6.Program.Method2(Type)' does not have matching annotations. The
+source value must declare at least the same requirements as those declared on the target
+location it is assigned to.
+```
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:162]: I would use a blockquote for this.

File: docs/core/deploying/fixing-warnings.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+ms.author: angocke
+ms.date: 09/08/2021
+---
+# Introduction to trim warnings
+
+Conceptually, trimming is very simple: when publishing the application, the .NET SDK analyzes the
+entire application and removes all unused code. However, it can be difficult to determine what is
+unused, or more precisely, what is used.
+
+To prevent changes in behavior when trimming applications, the .NET SDK provides static analysis of
```
Comment (adegeo) [inline docs/core/deploying/fixing-warnings.md:14]: You should link to other publishing/trimming articles in the intro

**Review by adegeo — CHANGES_REQUESTED** [review]
Thank you for submitting this article, there are a few things to handle as well as adding this article to the TOC and linking to it from other related publishing articles.

**Review by sbomer — COMMENTED** [review]
This reads really nicely! It needs to be added to the TOC at https://github.com/dotnet/docs/blob/3e6c9e8767e53ad5b7cf90c975bf70900b745380/docs/fundamentals/toc.yml#L1576

I'm wondering how to deal with the overlap with https://docs.microsoft.com/en-us/dotnet/core/deploying/prepare-libraries-for-trimming#resolve-trim-warnings. I think your doc is a better landing page and intro to the warnings. Maybe we could separate things out into the following?

- Prepare .NET libraries for trimming, landing page for library authors
  - move the section "Resolving trim warnings" out to a separate page (intended to provide the next level of detail compared to this intro)
  - link to this doc instead

- This doc, landing page for the trim warnings
  - at the end of the walkthrough, link to the separate doc about "Resolving trim warnings"

**Comment by agocke** [comment]
@sbomer Yup, there's a lot of overlap here. My approach was to try to get this in first, then I was going to try to consolidate and get your review. Happy to include whatever you think is useful

**Review by adegeo — APPROVED** [review]
Looks great!

**Comment by agocke** [comment]
Thanks! I'll try to follow-up later by removing duplication in other files and centralizing links.

**Pull request merged**